### PR TITLE
Release purview microsoft.purview preview/2024 04 01 preview

### PR DESCRIPTION
**Pull request by dwilfert-microsoft**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://aka.ms/brch

### CODE REVIEW

**Comment by ddumesnil-microsoft** [comment]
Regarding the LintDiff warning about boolean vs enum, I believe for the proposed field, a boolean should work as it is an irreversible state and only intended to be true or false. It is also scoped inside of the containing object as that is the only scenario it is intended to be set in and is for a specific use case and not a generalized use case.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: I am curious about whether you intend to allow setting it to the NotSpecified state, and whether that is what it defaults to - you might want to update the description to clarify

Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: Now that I see it, I also feel a concern that you allow setting publicNetworkAccess to NotSpecified and that I'm not familiar with any other RPs doing that for private links Is that a 'new' invention from your RP?

Or am I just out of date with my knowledge of what network team recommended for this?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1726]: How can it be used to set it, if its readonly?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1677]: Suggested rewording: 'This account's type, for the merge operation.'

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1659]: Did you see the warning about how booleans are not recommended because they're not extensible? They also tend to lead to you having to combine many booleans, in order to express more states, when potentially one enum could have done the job.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1637]: [ARMBlocking] how are you going to do RBAC permissions check on this?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1633]: Will you be adding any new examples that will show what this looks like? I'm still trying to understand the scenario, and noticed I couldn't find any.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2700]: It seems like you're changing your private links related definitions. Would this be a good opportunity to refactor to using common-types privatelinks.json?

Any blockers to doing so that you've identified?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: “Not specified” is for our internal logic to differentiate between existing accounts vs newly created accounts. But once the call is received, it’ll be set to enabled/disabled based on the internal checks.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1726]: Updated the description to only state Gets the default domain. Also added into the Accounts_Get example.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json]: Updated the wording.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1659]: Yes, I did see the warning. For this case, a boolean should work as it is an irreversible state and only intended to be true or false. It is also scoped inside of the containing object as that is the only scenario it is intended to be set in and is for a specific use case and not a generalized use case.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1637]: This is purely for reporting and informational purposes. Calls will not be performed on the account listed inside of the Account Merge Info.
To overshare, this object contains information from us merging two Purview accounts data. The account location, name, resource group, and subscription id properties are for us to know which account this account merged with. It is solely for informational purposes.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1633]: Ah, my apologies! I have added this field to the Accounts_Get.json example file. The scenario is also described in a different comment. Hope this helps! :)

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2700]: I looked into the common-types and have a few questions that could potentially present as blockers. There are some fields that seem to misalign from what we store in our repository and the common-types definitions.
For example, common-types defines PrivateEndpointConnection as a Resource, but we represent it as a ProxyResource. Similarly, the group-ids field in PrivateEndpointConnectionProperties is not present in our repo (likely because we are only storing the information as a proxy and connection, not the full resource) so that also misaligns.

What is the advice here in terms of refactoring? My concern on switching to the common-types is we may lose this nuance in our system (e.g. ProxyResource) and have data that is never actually returned by our system since we do not store it anywhere (e.g. group-ids).

**Comment by TimLovellSmith** [comment]
>           "x-ms-client-flatten": true

FWIW I think adding x-ms-flatten is sdk breaking, and not really considered 'better practice' these days. I don't think it was really worth it just to fix linter warnings, in case that is the only reason its being added

---
Refers to: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2701 in ecf736f. [](commit_id = ecf736f5936839b8ca4c48d5541a92c12a35954a, deletion_comment = False)

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: So, do I understand correctly that you would never actually want 'NotSpecified' to be sent in the request, and never want to return it in the response?
If I got that right, it seems quite confusing to add it to the enumeration!

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2700]: I think technically `ProxyResource` should be correct, and they are both similar - both definitions should not include things that are specific to TrackedResource. I don't think switching between them is likely to cause any issues, and the automated breaking changes validations that run on PRs here should hopefully catch any issues.

I am not sure whether group-ids is one of those parts of the private link contract that is not universally the same between RPs. I'd suggest we try to clarify that with the owners of the contract.

As long as the property is optional, and not required, I think technically its fine not to return it - see, for instance the way nextLink is always declared, but not actually always in responses. Adding it to your documented properties may look like a breaking change, and we can run that past breaking change reviewers to see what they think about the issue, but if you never actually return the property, or never return it except for new api-versions, I don't see how adding it to your documented set of optional properties is going to ever break anybody in practice.

I think it would still be an SDK breaking change, but hopefully if so it is at least one with some small upsides - overall helping better structure and maintain SDKs - and you're wanting to make some changes to private links to fix quality issues in your definitions anyway right? Including one enum change: PrivateEndpointConnectionStatus

Would breaking change experts like to add more guidance on whether its better or not to refactor at this point?

/cc @mikekistler @JeffreyRichter

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: - If the value is explicitly given as 'Disabled'/'Enabled' - we honor it.
- If it is 'NotSpecified' - 
  - If it is a new account creation after the feature launch- it will be enabled by default.
  - If it is updating an exisiting account - it'll be disabled by default till the customer performs mandatory setup before enabling it.
  - If a value is continued to be 'NotSpecified'- it helps us differentiate that its an existing account that did not get updated after the feature launch.

**Comment by JeffreyRichter** [comment]
Adding a new optional property to an API contract is not breaking; however, it still requires a new service api-version so customers can know what api-version supports the new property.

**Review by TimLovellSmith — CHANGES_REQUESTED** [review]
For your schema definitions, please reference the common-types schema definitions from (some version of) privatelinks.json instead of adding your own. If you face any blockers with that please let me know.

e.g.

```
          "200": {
            "description": "OK - Returns the specified of private link resource",
            "schema": {
              "$ref": "../../../../../common-types/resource-management/v6/privatelinks.json#/definitions/PrivateLinkResource"
            }
          },
```

**Comment by rkmanda** [comment]
Please add a workitem to refactor the private links related definitions.

**Comment by ddumesnil-microsoft** [comment]
@rkmanda here is the task to update to common-types for Private Links: [Task 3198663](https://msdata.visualstudio.com/Babylon/_workitems/edit/3198663): Update Private Links/Connections/Endpoints to use Common Types for GA.
Would you be able to remove the Change Request from Tim to unblock merging or was that intentional?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: @ddumesnil-microsoft
But why do you need a 'NotSpecified' value at all, in json format? Why not just have the field be modeled as optional? 
Since you seem to say it won't return NotSpecified in the response. It will return Disabled or Enabled, is that right?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1659]: I get that you're thinking of deprovisioning as a one-time irreversible thing, but is it really something that happens independently and orthogonally of all the other state transitions your resource can go through?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1637]: But how, for instance, do you avoid spoofing?

Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1637]: (BTW thanks for clarifying that these properties are mostly readonly, maybe that answers the concern?)

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1726]: BTW in case it leads to a doc clarification - is this a DNS domain? FQDN? Or something else?

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1791]: IMO the doc could probably be more verbose to provide more context

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [inline specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2978]: Sometimes people remember to annotate these as "format": "uri". I'm not sure how much it matters, TBH.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1637]: Yes, through this API, they will only be read only. This information is only exposed to users within the same tenant where they would also have knowledge of this account information. I don't believe there should be any concerns with these fields as they are only for information. So when an account is loaded after a merge, the UX can display which account it was merged to.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1659]: Yes, I totally understand your concerns, but in this specific case, it is intended to be one-time and irreversible. This field is merely an indicator of the account state and is intended to be transitional as well. Through this API, we will use this field to display some UX messages to a user. It is similar to a deleted state and is a one-way operation.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1840]: NotSpecified could be returned in the response. It is the value that is used for accounts that existed before the feature launched (as new accounts are Enabled by default and could otherwise be updated to Disabled) and have not undergone any updates (as that would set it to Disabled by default).

This allows us to have easier tracked values as the value will be present for all accounts rather than using `null` as the NotSpecified value.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1726]: It is something else. Domain in this case is an internal term for a Root layer Collection of Objects (in Purview)

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:1791]: I will add this into the above work item, but would like to avoid pushing additional changes at this time if not required.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (TimLovellSmith) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2700]: PS my general understanding, is that group-id should be specified by your service since it used when creating private endpoint connection(s) to specify the logical 'group' of endpoints which this private endpoint connection is to connect to. For instance, some services might have multiple hosted-on-behalf FQDNs that the customer can use, to connect to the service, and choose to split those up into functional groups for specific purposes the customer wishes to enable, e.g. (data ingestion, service management, persistence, monitoring).

I can see that maybe your API seems to be fine without providing this kind of information, but I also feel that in principal all private endpoint connections should provide this information. And it shouldn't actually be a breaking change, to start returning it, when previously or not.

**Review by TimLovellSmith — APPROVED** [review]
Unblocking for now, with the expectation that you'll follow up with Private Link experts on the pros and cons of adding the group id property, or any other missing properties, in your response.

File: specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json
Comment (ddumesnil-microsoft) [reply specification/purview/resource-manager/Microsoft.Purview/preview/2024-04-01-preview/purview.json:2700]: I will add this into the task to also investigate the addition of group ids for private endpoints. Thanks for the information Tim!

**Pull request merged**
